tests/src/Kernel/VocabularyReferenceFormatterTest.php

```diff
@@ -0,0 +1,400 @@
+      'tags' => ['config:open_vocabularies.open_vocabulary_association.test_entities.related'],
+      'max-age' => Cache::PERMANENT,
+    ], $build[2]['#cache']);
+
+    // Each item list has the cache metadata of the referenced entity.
+    $this->assertEquals([
+      'contexts' => [],
+      'tags' => ['entity_test_with_bundle:1'],
+      'max-age' => Cache::PERMANENT,
+    ], $build[0]['#items'][0]['#cache']);
+    $this->assertEquals([
+      'contexts' => [],
+      'tags' => ['entity_test_with_bundle:3'],
+      'max-age' => Cache::PERMANENT,
+    ], $build[0]['#items'][1]['#cache']);
+    $this->assertEquals([
+      'contexts' => [],
+      'tags' => ['entity_test_with_bundle:2'],
+      'max-age' => Cache::PERMANENT,
+    ], $build[0]['#items'][2]['#cache']);
+    $this->assertEquals([
+      'contexts' => [],
+      'tags' => ['entity_test_with_bundle:1'],
+      'max-age' => Cache::PERMANENT,
+    ], $build[1]['#items'][0]['#cache']);
+    $this->assertEquals([
+      'contexts' => [],
+      'tags' => ['entity_test:1'],
+      'max-age' => Cache::PERMANENT,
+    ], $build[2]['#items'][0]['#cache']);
+
+    // Render the formatter output.
+    $html = (string) \Drupal::service('renderer')->renderRoot($build);
+    $crawler = new Crawler($html);
+    $wrapper = $crawler->filter('div.field__items');
+    $this->assertCount(1, $wrapper);
+    $lists = $wrapper->filter('div.field__item div.item-list');
+    $this->assertCount(3, $lists);
+
+    // Assert the titles of each list.
+    $this->assertEquals('Participating countries', $lists->eq(0)->filter('h3')->html());
+    $this->assertEquals('Hosting country', $lists->eq(1)->filter('h3')->html());
+    $this->assertEquals('Related entities', $lists->eq(2)->filter('h3')->html());
+
+    // Verify the items in the first list.
+    $items = $lists->eq(0)->filter('ul li');
+    $this->assertCount(3, $items);
+    $this->assertEquals('Italy', $items->eq(0)->html());
+    $this->assertEquals('Lithuania', $items->eq(1)->html());
+    $this->assertEquals('Belgium', $items->eq(2)->html());
+    // And in the second.
+    $items = $lists->eq(1)->filter('ul li');
+    $this->assertCount(1, $items);
+    $this->assertEquals('Italy', $items->html());
+    // Amd finally the third.
+    $items = $lists->eq(2)->filter('ul li');
+    $this->assertCount(1, $items);
+    $this->assertEquals('Referenced', $items->html());
+
+    // Set the formatter to link to the entity.
+    \Drupal::service('entity_display.repository')
+      ->getViewDisplay('entity_test', 'entity_test')
+      ->setComponent('field_test', [
+        'type' => 'open_vocabulary_reference_label',
+        'settings' => [
+          'link' => TRUE,
+        ],
+      ])
+      ->save();
+
+    $build = $entity->get('field_test')->view('default');
+    $html = (string) \Drupal::service('renderer')->renderRoot($build);
+    $crawler = new Crawler($html);
+
+    // Verify that the labels are linked to the entity path.
```

Comment:
Can we test the exception and that will fall back to non-linked rendering?

Reply:
I left out quite a few negative cases as testing for them is incredibly hard unfortunately. How do I make an entity return an exception in that case? I also would have had to mock most of the services to check all the instances where I bail out if no vocabulary/handler/association are found.